File: src/recency.py
"""Hard recency rules: only jobs posted in the last 7 or 8–14 days."""
from __future__ import annotations

import re
from datetime import date
from typing import Any

from models import Job

MAX_AGE_DAYS = 14
WEEK_1_DAYS = 7

_MONTHS = {
    "january": 1,
    "jan": 1,
    "february": 2,
    "feb": 2,
    "march": 3,
    "mar": 3,
    "april": 4,
    "apr": 4,
    "may": 5,
    "june": 6,
    "jun": 6,
    "july": 7,
    "jul": 7,
    "august": 8,
    "aug": 8,
    "september": 9,
    "sep": 9,
    "sept": 9,
    "october": 10,
    "oct": 10,
    "november": 11,
    "nov": 11,
    "december": 12,
    "dec": 12,
}


def parse_posted_age_days(text: str, *, today: date | None = None) -> tuple[int | None, str]:
    """Return (age_in_days, evidence) if a posting date/relative age is found."""
    today = today or date.today()
    t = (text or "").lower().replace("–", "-").replace("—", "-")

    for m in re.finditer(r"\b(20\d{2})[-/](\d{1,2})[-/](\d{1,2})\b", t):
        try:
            d = date(int(m.group(1)), int(m.group(2)), int(m.group(3)))
            return (today - d).days, m.group(0)
        except ValueError:
            pass

    for m in re.finditer(
        r"\b(?:posted\s+(?:on\s+)?)?(january|february|march|april|may|june|july|august|"
        r"september|october|november|december|jan|feb|mar|apr|jun|jul|aug|sep|sept|oct|nov|dec)"
        r"\s+(\d{1,2})(?:st|nd|rd|th)?(?:,?\s*(20\d{2}))?\b",
        t,
    ):
        mon = _MONTHS.get(m.group(1)) or _MONTHS.get(m.group(1)[:3])
        if not mon:
            continue
        day_n = int(m.group(2))
        year = int(m.group(3)) if m.group(3) else today.year
        try:
            d = date(year, mon, day_n)
        except ValueError:
            continue
        if not m.group(3) and d > today:
            try:
                d = date(year - 1, mon, day_n)
            except ValueError:
                continue
        return (today - d).days, m.group(0)

    for m in re.finditer(
        r"\b(\d{1,2})\s+(january|february|march|april|may|june|july|august|"
        r"september|october|november|december|jan|feb|mar|apr|jun|jul|aug|sep|sept|oct|nov|dec)"
        r"(?:,?\s*(20\d{2}))?\b",
        t,
    ):
        mon = _MONTHS.get(m.group(2)) or _MONTHS.get(m.group(2)[:3])
        if not mon:
            continue
        day_n = int(m.group(1))
        year = int(m.group(3)) if m.group(3) else today.year
        try:
            d = date(year, mon, day_n)
        except ValueError:
            continue
        if not m.group(3) and d > today:
            try:
                d = date(year - 1, mon, day_n)
            except ValueError:
                continue
        return (today - d).days, m.group(0)

    # API epoch seconds / millis in posted_at alone
    if re.fullmatch(r"\d{10}", text.strip()):
        try:
            from datetime import datetime, timezone

            d = datetime.fromtimestamp(int(text.strip()), tz=timezone.utc).date()
            return (today - d).days, f"epoch:{text.strip()}"
        except (ValueError, OSError):
            pass
    if re.fullmatch(r"\d{13}", text.strip()):
        try:
            from datetime import datetime, timezone

            d = datetime.fromtimestamp(int(text.strip()) / 1000, tz=timezone.utc).date()
            return (today - d).days, f"epoch_ms:{text.strip()}"
        except (ValueError, OSError):
            pass

    if re.search(r"\b(just posted|posted today|today|hours?\s+ago|minutes?\s+ago)\b", t):
        return 0, "today/hours-ago"
    if re.search(r"\byesterday\b", t):
        return 1, "yesterday"
    m = re.search(r"\b(\d+)\s*hours?\s+ago\b", t)
    if m:
        return 0, m.group(0)
    m = re.search(r"\b(\d+)\s*days?\s+ago\b", t)
    if m:
        return int(m.group(1)), m.group(0)
    m = re.search(r"\b(\d+)\s*weeks?\s+ago\b", t)
    if m:
        return int(m.group(1)) * 7, m.group(0)
    m = re.search(r"\b(\d+)\s*months?\s+ago\b", t)
    if m:
        return int(m.group(1)) * 30, m.group(0)
    m = re.search(r"\b(\d+)\s*years?\s+ago\b", t)
    if m:
        return int(m.group(1)) * 365, m.group(0)
    if re.search(r"\ba\s+year\s+ago\b", t):
        return 365, "a year ago"
    if re.search(r"\blast\s+month\b", t):
        return 30, "last month"
    if re.search(r"\bthis\s+week\b", t):
        return 3, "this week"
    if re.search(r"\blast\s+week\b", t):
        return 7, "last week"

    return None, ""


def check_recency(
    job: Job | dict[str, Any],
    *,
    max_age_days: int = MAX_AGE_DAYS,
    reject_unknown: bool = True,
    week_1_days: int = WEEK_1_DAYS,
    today: date | None = None,
) -> tuple[bool, str, int | None, str]:
    """Return (ok, reason, age_days, bucket)."""
    if isinstance(job, Job):
        blob = " ".join(
            [
                job.posted_at or "",
                job.title or "",
                job.summary or "",
                job.location or "",
            ]
        )
        if job.posted_at and job.posted_at.isdigit():
            age, evidence = parse_posted_age_days(job.posted_at, today=today)
            if age is None:
                age, evidence = parse_posted_age_days(blob, today=today)
        else:
            age, evidence = parse_posted_age_days(blob, today=today)
    else:
        blob = " ".join(
            str(job.get(k) or "")
            for k in ("posted_at", "date", "published", "title", "summary", "location")
        )
        age, evidence = parse_posted_age_days(blob, today=today)

    if age is None:
        if reject_unknown:
            return False, "REJECT: no posting date (unknown/missing)", None, "unknown"
        return True, "posted_at:unknown (allowed)", None, "unknown"
    if age < 0:
        return False, f"REJECT: invalid/future posting date ({evidence})", age, "unknown"
    if age > max_age_days:
        return (
            False,
            f"REJECT: older than {max_age_days}d ({evidence}, ~{age}d)",
            age,
            "stale",
        )
    bucket = "last_7_days" if age <= week_1_days else "days_8_to_14"
    return True, f"fresh:~{age}d ({evidence})", age, bucket


def apply_recency(
    job: Job,
    *,
    max_age_days: int = MAX_AGE_DAYS,
    reject_unknown: bool = True,
    week_1_days: int = WEEK_1_DAYS,
) -> Job:
    ok, reason, age, bucket = check_recency(
        job,
        max_age_days=max_age_days,
        reject_unknown=reject_unknown,
        week_1_days=week_1_days,
    )
    job.posted_age_days = age
    job.recency_bucket = bucket
    if not ok:
        job.eligible = False
        job.score = 0
        job.reject_reason = reason
    return job

File: src/lib/recency.ts
import type { Job, RecencyBucket } from "./types";
import { DEFAULT_RECENCY } from "./config";

const MONTHS: Record<string, number> = {
  january: 1,
  jan: 1,
  february: 2,
  feb: 2,
  march: 3,
  mar: 3,
  april: 4,
  apr: 4,
  may: 5,
  june: 6,
  jun: 6,
  july: 7,
  jul: 7,
  august: 8,
  aug: 8,
  september: 9,
  sep: 9,
  sept: 9,
  october: 10,
  oct: 10,
  november: 11,
  nov: 11,
  december: 12,
  dec: 12,
};

function todayUTC(): Date {
  const now = new Date();
  return new Date(
    Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate()),
  );
}

function daysBetween(a: Date, b: Date): number {
  return Math.floor((a.getTime() - b.getTime()) / 86_400_000);
}

export function parsePostedAgeDays(
  text: string,
  today: Date = todayUTC(),
): { age: number | null; evidence: string } {
  const t = (text || "").toLowerCase().replace(/[–—]/g, "-");

  let m = t.match(/\b(20\d{2})[-/](\d{1,2})[-/](\d{1,2})\b/);
  if (m) {
    const d = new Date(Date.UTC(+m[1], +m[2] - 1, +m[3]));
    if (!Number.isNaN(d.getTime())) {
      return { age: daysBetween(today, d), evidence: m[0] };
    }
  }

  const monthDay = t.match(
    /\b(?:posted\s+(?:on\s+)?)?(january|february|march|april|may|june|july|august|september|october|november|december|jan|feb|mar|apr|jun|jul|aug|sep|sept|oct|nov|dec)\s+(\d{1,2})(?:st|nd|rd|th)?(?:,?\s*(20\d{2}))?\b/,
  );
  if (monthDay) {
    const mon = MONTHS[monthDay[1]] ?? MONTHS[monthDay[1].slice(0, 3)];
    if (mon) {
      let year = monthDay[3] ? +monthDay[3] : today.getUTCFullYear();
      let d = new Date(Date.UTC(year, mon - 1, +monthDay[2]));
      if (!monthDay[3] && d > today) {
        d = new Date(Date.UTC(year - 1, mon - 1, +monthDay[2]));
      }
      if (!Number.isNaN(d.getTime())) {
        return { age: daysBetween(today, d), evidence: monthDay[0] };
      }
    }
  }

  if (/^\d{10}$/.test(text.trim())) {
    const d = new Date(+text.trim() * 1000);
    if (!Number.isNaN(d.getTime())) {
      return {
        age: daysBetween(
          today,
          new Date(
            Date.UTC(d.getUTCFullYear(), d.getUTCMonth(), d.getUTCDate()),
          ),
        ),
        evidence: `epoch:${text.trim()}`,
      };
    }
  }
  if (/^\d{13}$/.test(text.trim())) {
    const d = new Date(+text.trim());
    if (!Number.isNaN(d.getTime())) {
      return {
        age: daysBetween(
          today,
          new Date(
            Date.UTC(d.getUTCFullYear(), d.getUTCMonth(), d.getUTCDate()),
          ),
        ),
        evidence: `epoch_ms:${text.trim()}`,
      };
    }
  }

  if (
    /\b(just posted|posted today|today|hours?\s+ago|minutes?\s+ago)\b/.test(t)
  ) {
    return { age: 0, evidence: "today/hours-ago" };
  }
  if (/\byesterday\b/.test(t)) return { age: 1, evidence: "yesterday" };
  m = t.match(/\b(\d+)\s*hours?\s+ago\b/);
  if (m) return { age: 0, evidence: m[0] };
  m = t.match(/\b(\d+)\s*days?\s+ago\b/);
  if (m) return { age: +m[1], evidence: m[0] };
  m = t.match(/\b(\d+)\s*weeks?\s+ago\b/);
  if (m) return { age: +m[1] * 7, evidence: m[0] };
  m = t.match(/\b(\d+)\s*months?\s+ago\b/);
  if (m) return { age: +m[1] * 30, evidence: m[0] };
  if (/\bthis\s+week\b/.test(t)) return { age: 3, evidence: "this week" };
  if (/\blast\s+week\b/.test(t)) return { age: 7, evidence: "last week" };

  return { age: null, evidence: "" };
}

function bucketForAge(age: number | null, week1: number): RecencyBucket {
  if (age === null) return "unknown";
  if (age < 0) return "unknown";
  if (age <= week1) return "last_7_days";
  if (age <= 14) return "days_8_to_14";
  if (age <= 21) return "days_15_to_21";
  if (age <= 30) return "days_22_to_30";
  return "older";
}

/**
 * Annotate age/bucket only — never reject. Age filters belong in the UI.
 */
export function annotateRecency(
  job: Job,
  opts?: { week_1_days?: number },
): Job {
  const week1 = opts?.week_1_days ?? DEFAULT_RECENCY.bucket_week_1_days;
  const blob = [job.posted_at, job.title, job.summary, job.location]
    .filter(Boolean)
    .join(" ");

  let parsed = job.posted_at
    ? parsePostedAgeDays(job.posted_at)
    : { age: null as number | null, evidence: "" };
  if (parsed.age === null) {
    parsed = parsePostedAgeDays(blob);
  }

  job.posted_age_days = parsed.age;
  job.recency_bucket = bucketForAge(parsed.age, week1);
  return job;
}

/** @deprecated use annotateRecency — kept for tests that check age parsing */
export function checkRecency(
  job: Job,
  opts?: {
    max_age_days?: number;
    reject_unknown?: boolean;
    week_1_days?: number;
  },
): { ok: boolean; reason: string; age: number | null; bucket: RecencyBucket } {
  annotateRecency(job, { week_1_days: opts?.week_1_days });
  const age = job.posted_age_days ?? null;
  const bucket = job.recency_bucket || "unknown";
  const maxAge = opts?.max_age_days;
  const rejectUnknown = opts?.reject_unknown ?? false;

  if (age === null) {
    if (rejectUnknown) {
      return {
        ok: false,
        reason: "REJECT: no posting date (unknown/missing)",
        age: null,
        bucket: "unknown",
      };
    }
    return {
      ok: true,
      reason: "posted_at:unknown (allowed)",
      age: null,
      bucket: "unknown",
    };
  }
  if (maxAge != null && age > maxAge) {
    return {
      ok: false,
      reason: `REJECT: older than ${maxAge}d (~${age}d)`,
      age,
      bucket,
    };
  }
  return { ok: true, reason: `age:~${age}d`, age, bucket };
}

export function applyRecency(job: Job, profileRecency?: ProfileRecency): Job {
  // Web product: annotate only — do not hard-reject by age.
  return annotateRecency(job, {
    week_1_days: profileRecency?.bucket_week_1_days,
  });
}

type ProfileRecency = {
  max_age_days?: number;
  reject_unknown_date?: boolean;
  bucket_week_1_days?: number;
};
